### PR TITLE
Implement m2m-token for Trigger Command Execution via m2m-oauth-server

### PR DESCRIPTION
**Pull request by jkralik**
## Description

This PR introduces the `m2m-token` functionality, enabling the execution of trigger commands from various services. The `m2m-token` is issued by the `m2m-oauth-server` through the credential flow.

To obtain the token, users can utilize a secret or another JWT token signed by an authority configured within the `m2m-oauth-server`.

Key highlights of this implementation include:
- Integration with `m2m-oauth-server` for token issuance.
- Support for credential flow to obtain tokens.
- Flexibility to use a secret or another JWT token for authentication.

### CODE REVIEW

File: m2m-oauth-server/Makefile [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+build: build-servicecontainer
+
+
+push: build-servicecontainer
+	docker push plgd/$(SERVICE_NAME):$(VERSION_TAG)
+	docker push plgd/$(SERVICE_NAME):$(LATEST_TAG)
+
+proto/generate:
+.PHONY: proto/generate
```
Comment (Danielius1922) [inline m2m-oauth-server/Makefile]: duplicate with line 47

File: test/config/config.go [hunk 2]
```diff
@@ -43,6 +43,8 @@ const (
 	RESOURCE_DIRECTORY_HOST         = "localhost:20004"
 	CERTIFICATE_AUTHORITY_HOST      = "localhost:20011"
 	CERTIFICATE_AUTHORITY_HTTP_HOST = "localhost:20012"
+	// M2M_OAUTH_SERVER_HOST           = "localhost:20013" // for grpc
```
Comment (Danielius1922) [inline test/config/config.go]: rm?

File: pkg/security/jwt/validator/validator.go [hunk 3]
```diff
@@ -26,42 +27,54 @@ type Validator struct {
...
-	if err != nil {
-		return nil, fmt.Errorf("cannot create cert manager: %w", err)
-	}
+	keys := jwtValidator.NewMultiKeyCache()
+	var onClose fn.FuncList
+	openIDConfigurations := make([]openid.Config, 0, len(config.Endpoints))
+	for _, authority := range config.Endpoints {
+		httpClient, err := client.New(authority.HTTP, fileWatcher, logger, tracerProvider)
+		if err != nil {
+			return nil, fmt.Errorf("cannot create cert manager: %w", err)
```
Comment (Danielius1922) [inline pkg/security/jwt/validator/validator.go]: is cert manager correct here?

File: m2m-oauth-server/service/httpApi.go [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+type RequestHandler struct {
+	config                  *Config
+	authRestriction         *cache.Cache[string, struct{}]
+	accessTokenKey          interface{}
+	accessTokenJwkKey       jwk.Key
+	refreshRestriction      *cache.Cache[string, struct{}]
+	privateKeyJWTValidators map[string]*validator.Validator
+}
+
+func createJwkKey(privateKey interface{}) (jwk.Key, error) {
```
Comment (Danielius1922) [inline m2m-oauth-server/service/httpApi.go]: Same as `createJwkKey` in test/oauth-server, maybe we can create a common methond somewhere in pkg?

File: m2m-oauth-server/service/jsonWriter.go [hunk 5]
```diff
@@ -0,0 +1,19 @@
+package service
+
+import (
+	"net/http"
+
+	"github.com/plgd-dev/go-coap/v3/message"
+	"github.com/plgd-dev/kit/v2/codec/json"
+)
+
+const contentTypeHeaderKey = "Content-Type"
```
Comment (Danielius1922) [inline m2m-oauth-server/service/jsonWriter.go]: `ContentTypeOptionsHeaderKey` from pkg/net/http

File: bundle/Dockerfile [hunk 6]
```diff
@@ -160,6 +168,8 @@ COPY --from=build /go/bin/nats-server /usr/local/bin/nats-server
 COPY --from=build /go/bin/nats /usr/local/bin/nats
 COPY --from=build /go/src/github.com/plgd-dev/hub/bundle/jetstream.json /configs/jetstream.json
 COPY --from=build /go/bin/oauth-server /usr/local/bin/oauth-server
+COPY --from=build /go/bin/m2m-oauth-server /usr/local/bin/m2m-oauth-server
```
Comment (Danielius1922) [inline bundle/Dockerfile:172]: I'd move lines 171-172 below line 173, so the `oauth-server/config.yaml` is just below the oauth-server binary

File: charts/plgd-hub/templates/m2m-oauth-server/deployment.yaml [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+                  openssl ecparam -name prime256v1 -genkey -noout -out {{ $.Values.m2moauthserver.oauthSigner.accessTokenKeyFile }}
+                  
+              fi
+              chmod 0644 {{ $.Values.m2moauthserver.oauthSigner.accessTokenKeyFile }}
+              EOF
+          volumeMounts:
+            - mountPath: /keys
+              name: keys
+      containers:
+        - name: {{ .Values.m2moauthserver.name }}
```
[the remark this answers is not in the document]
Comment (Danielius1922) [reply charts/plgd-hub/templates/m2m-oauth-server/deployment.yaml:42]: @jkralik can we add the limit values SC wants?

File: charts/plgd-hub/templates/m2m-oauth-server/ingress.yaml [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    {{- include "plgd-hub.tplvalues.render" ( dict "value" .Values.m2moauthserver.ingress.customAnnotations "context" $ ) | nindent 4 }}
+    {{- end }}
+spec:
+  tls:
+    - hosts:
+        - {{ include "plgd-hub.m2moauthserver.ingressDomain" . | quote }}
+      {{- if $.Values.global.enableWildCartCert }}
+      secretName: {{ include "plgd-hub.wildCardCertName" . | quote }}
+      {{- else }}
+      secretName: {{ include "plgd-hub.m2moauthserver.domainCertName" . | quote }}
```
Comment (Danielius1922) [inline charts/plgd-hub/templates/m2m-oauth-server/ingress.yaml:25]: use `$domainCrt` here or remove `$domainCrt`, it's not used anywhere else in this file

File: m2m-oauth-server/Makefile (hunk 1 above)
Comment (jkralik) [reply m2m-oauth-server/Makefile]: removed

File: test/config/config.go (hunk 2 above)
Comment (jkralik) [reply test/config/config.go]: removed

File: pkg/security/jwt/validator/validator.go (hunk 3 above)
Comment (jkralik) [reply pkg/security/jwt/validator/validator.go]: yes - added there client cert manager

File: m2m-oauth-server/service/httpApi.go (hunk 4 above)
Comment (jkralik) [reply m2m-oauth-server/service/httpApi.go]: moved to jwt package

File: m2m-oauth-server/service/jsonWriter.go (hunk 5 above)
Comment (jkralik) [reply m2m-oauth-server/service/jsonWriter.go]: fixed

File: bundle/Dockerfile (hunk 6 above)
Comment (jkralik) [reply bundle/Dockerfile:172]: fixed

File: charts/plgd-hub/templates/m2m-oauth-server/ingress.yaml (hunk 8 above)
Comment (jkralik) [reply charts/plgd-hub/templates/m2m-oauth-server/ingress.yaml:25]: removed

**Comment by jkralik** [comment]
https://github.com/plgd-dev/hub/pull/1336#discussion_r1663991640

This depends on the cluster where the hub is running. In production, each user needs to set up their own cluster. If we set the limits too low, the processes will be killed by OOM (Out of Memory) errors.

**Comment by jkralik** [comment]
@coderabbitai review

**Comment by jkralik** [comment]
@coderabbitai summary

File: cloud2cloud-gateway/service/retrieveAllDevicesSubscription_test.go [hunk 9]
```diff
@@ -101,7 +101,7 @@ func TestRequestHandlerRetrieveDevicesSubscription(t *testing.T) {
 			},
 			wantCode:        http.StatusUnauthorized,
 			wantContentType: textPlain,
-			want:            "invalid token: could not parse token: token has invalid claims: inaccessible uri",
+			want:            "unknown issuer",
```
Comment (Danielius1922) [inline cloud2cloud-gateway/service/retrieveAllDevicesSubscription_test.go]: we lose the real error, which remains `inaccessbile uri` from `DeniedClaims`. We should at least make an error log in DeniedClaims with the error is created, so we know that this occured. 

`func (c *MultiKeyCache) GetOrFetchKeyWithContext` gets called before `func (c DeniedClaims) Validate()` by the external jwt/v5 package. I guess we could try casting the token.Claims in `GetOrFetchKeyWithContext` to DeniedClaims and return the error if it's not nil at the begining of the function. Though this is a maybe a bit of an overkill, the error log would be sufficient for me.

File: m2m-oauth-server/test/test.go [hunk 10]
```diff
@@ -0,0 +1,356 @@
...
+		body:        body,
+		uri:         fmt.Sprintf("https://%s%s", host, url),
+		uriParams:   make(map[string]interface{}),
+		header:      make(map[string]string),
+		queryParams: make(map[string]string),
+	}
+	return &b
+}
+
+type RequestBuilder struct {
```
Comment (Danielius1922) [inline m2m-oauth-server/test/test.go]: This will be the 4th RequestBuilder in the repository, can't the most general one in test/http be used here?

Comment (jkralik) [reply m2m-oauth-server/test/test.go]: using test/http

File: cloud2cloud-gateway/service/retrieveAllDevicesSubscription_test.go (hunk 9 above)
Comment (jkralik) [reply cloud2cloud-gateway/service/retrieveAllDevicesSubscription_test.go]: fixed -> if is Error interface then it will return directly the error

Comment (Danielius1922) [reply cloud2cloud-gateway/service/retrieveAllDevicesSubscription_test.go]: 👍

File: m2m-oauth-server/test/test.go (hunk 10 above)
Comment (Danielius1922) [reply m2m-oauth-server/test/test.go]: rm?

Comment (jkralik) [reply m2m-oauth-server/test/test.go]: removed

**Review by Danielius1922 — APPROVED** [review]

**Comment by jkralik** [comment]
@coderabbitai summary

**Comment by jkralik** [comment]
@coderabbitai create summary over files under m2m-oauth-server directory and helm chart

**Comment by jkralik** [comment]
@coderabbitai create summary over all files under m2m-oauth-server and charts directories

**Comment by jkralik** [comment]
@coderabbitai create a description for whole pull request based on `m2m-oauth-server` and `charts`  directories

**Pull request merged**